src/python_be.cc

```diff
@@ -661,46 +660,33 @@ ModelInstanceState::ExecuteBLSRequest(
             lbackend_memory.reset(backend_memory);
             input_tensor->SetMemory(std::move(PbMemory::Create(
                 Stub()->ShmPool(), std::move(lbackend_memory))));
+            gpu_buffer_transporter.AddBuffer(
+                input_tensor->Memory()->ShmHandle());
 #endif  // TRITON_ENABLE_GPU
           }
         }
       }
       catch (const PythonBackendException& exception) {
+        gpu_buffer_transporter.SetError(Stub()->ShmPool(), exception.what());
         pb_exception = exception;
       }
-      AllocatedSharedMemory<bi::managed_external_buffer::handle_t> gpu_handles;
 
       // Wait for the extra round trip to complete. The stub process will fill
       // in the data for the GPU tensors. If there is an error, the extra round
       // trip must be still completed, otherwise the stub process will always be
       // waiting for a message from the parent process.
       if (has_gpu_tensor) {
-        try {
-          gpu_handles = Stub()
-                            ->ShmPool()
-                            ->Construct<bi::managed_external_buffer::handle_t>(
-                                gpu_buffers_count);
-          request_batch_shm_ptr->gpu_buffers_count = gpu_buffers_count;
-          request_batch_shm_ptr->gpu_buffers_handle = gpu_handles.handle_;
-          size_t i = 0;
-          for (auto& input_tensor : infer_request->Inputs()) {
-            if (!input_tensor->IsCPU()) {
-              gpu_handles.data_.get()[i] = input_tensor->Memory()->ShmHandle();
-              ++i;
-            }
-          }
-        }
-        catch (const PythonBackendException& exception) {
-          pb_exception = exception;
-        }
+        gpu_buffer_transporter.Complete(Stub()->ShmPool());
```

Comment:
we pass in the shmpool to complete and seterror - could this be added to the constructor and stored ? Does the pool change for the same transporter object?

Reply:
ShmPool is a reference to a unique pointer. I think the lifetime of a reference to a unique pointer stored as a class attribute is a bit weird as the object can continue to use the attribute after it is lifetime is finished which would result in segfault. I think this is more clear in terms of lifetime. Alternatively, we could return a shared pointer but that is slightly less performant than what we have here (it would have to obtain a lock, increase/decrease ref count for shared pointer).

Reply:
I agree that we should be careful about managing the lifetime. Passing ShmPool instead of storing it looks clearer to me as well. It seems like we do store a reference to a unique pointer in some of the objects though, [ResponseSender](https://github.com/triton-inference-server/python_backend/blob/a156eba0cc03739a5bdcf59da545d00c531e4ee3/src/response_sender.h#L44) for example.